clippy_lints/src/await_holding_lock.rs

```diff
@@ -0,0 +1,100 @@
+use crate::utils::span_lint_and_note;
+use if_chain::if_chain;
+use rustc_hir::intravisit::FnKind;
+use rustc_hir::{Body, FnDecl, HirId, IsAsync};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::{Span, Symbol};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for calls to await while holding a MutexGuard.
+    ///
+    /// **Why is this bad?** This is almost certainly an error which can result
+    /// in a deadlock because the reactor will invoke code not visible to the
+    /// currently visible scope.
+    ///
+    /// **Known problems:** Detects only specifically named guard types:
+    /// MutexGuard, RwLockReadGuard, and RwLockWriteGuard.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// use std::sync::Mutex;
+    ///
+    /// async fn foo(x: &Mutex<u32>) {
+    ///   let guard = x.lock().unwrap();
+    ///   *guard += 1;
+    ///   bar.await;
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::sync::Mutex;
+    ///
+    /// async fn foo(x: &Mutex<u32>) {
+    ///   {
+    ///     let guard = x.lock().unwrap();
+    ///     *guard += 1;
+    ///   }
+    ///   bar.await;
+    /// }
+    /// ```
+    pub AWAIT_HOLDING_LOCK,
+    pedantic,
+    "Inside an async function, holding a MutexGuard while calling await"
+}
+
+const MUTEX_GUARD_TYPES: [&str; 3] = ["MutexGuard", "RwLockReadGuard", "RwLockWriteGuard"];
+
+declare_lint_pass!(AwaitHoldingLock => [AWAIT_HOLDING_LOCK]);
+
+impl LateLintPass<'_, '_> for AwaitHoldingLock {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'_, '_>,
+        fn_kind: FnKind<'_>,
+        _: &FnDecl<'_>,
+        _: &Body<'_>,
+        span: Span,
+        _: HirId,
+    ) {
+        if !is_async_fn(fn_kind) {
+            return;
+        }
+
+        for ty_clause in &cx.tables.generator_interior_types {
```

Comment:
how does this work?

Reply:
My understanding based on the way check_fn gets called is that cx.tables gets set up here: https://github.com/rust-lang/rust/blob/58dd1ce8383aaebcad9b6027b89a316fd868b35c/src/librustc_lint/late.rs#L184

I am pretty sure these are the same tables as:

```
let def_id = cx.tcx.hir().local_def_id(hir_id);
let tables = cx.tcx.typeck_tables_of(def_id);
```

where `hir_id` is passed into check_fn and cx is the LateContext.

Okay so that explains the where the TypeckTables come from. Then generator_interior_types is a vector of GeneratorInteriorTypeCause. The docs for that type says

> Whenever a value may be live across a generator yield, the type of that value winds up in the GeneratorInteriorTypeCause struct.

So if a type ends up inside one of these TypeCause structs inside said vector then it is being held across a yield (which I am assuming is coming from an await).

Reading this now I realize I used the variable name "ty_clause" when I probably should have used "ty_cause".

Reply:
Ah, TIL `cx.tables` gets updated per function, it used to work differently